File: apollo-android-support/src/main/java/com/apollographql/apollo/ApolloPrefetchCallback.java
package com.apollographql.apollo;

import android.os.Handler;

import com.apollographql.apollo.exception.ApolloException;
import com.apollographql.apollo.exception.ApolloHttpException;
import com.apollographql.apollo.exception.ApolloNetworkException;

import javax.annotation.Nonnull;

import static com.apollographql.apollo.api.internal.Utils.checkNotNull;

/**
 * Android wrapper for {@link ApolloPrefetch.Callback} to be operated on specified {@link Handler}
 */
public final class ApolloPrefetchCallback extends ApolloPrefetch.Callback {
  final ApolloPrefetch.Callback delegate;
  private final Handler handler;

  /**
   * Wraps {@code callback} to be be operated on specified {@code handler}
   *
   * @param callback original callback to delegates calls
   * @param handler  the callback will be run on the thread to which this handler is attached
   */
  public static <T> ApolloPrefetchCallback wrap(@Nonnull ApolloPrefetch.Callback callback, @Nonnull Handler handler) {
    return new ApolloPrefetchCallback(callback, handler);
  }

  /**
   * @param callback original callback to delegates calls
   * @param handler  the callback will be run on the thread to which this handler is attached
   */
  public ApolloPrefetchCallback(@Nonnull ApolloPrefetch.Callback callback, @Nonnull Handler handler) {
    this.delegate = checkNotNull(callback, "callback == null");
    this.handler = checkNotNull(handler, "handler == null");
  }

  @Override public void onSuccess() {
    handler.post(new Runnable() {
      @Override public void run() {
        delegate.onSuccess();
      }
    });
  }

  @Override public void onFailure(@Nonnull final ApolloException e) {
    handler.post(new Runnable() {
      @Override public void run() {
        delegate.onFailure(e);
      }
    });
  }

  @Override public void onHttpError(@Nonnull final ApolloHttpException e) {
    handler.post(new Runnable() {
      @Override public void run() {
        delegate.onHttpError(e);
      }
    });
  }

  @Override public void onNetworkError(@Nonnull final ApolloNetworkException e) {
    handler.post(new Runnable() {
      @Override public void run() {
        delegate.onNetworkError(e);
      }
    });
  }
}

File: apollo-android-support/src/main/kotlin/com/apollographql/apollo/ApolloPrefetchCallback.kt
package com.apollographql.apollo

import android.os.Handler
import com.apollographql.apollo.api.internal.Utils.__checkNotNull
import com.apollographql.apollo.exception.ApolloException
import com.apollographql.apollo.exception.ApolloHttpException
import com.apollographql.apollo.exception.ApolloNetworkException

/**
 * Android wrapper for [ApolloPrefetch.Callback] to be operated on specified [Handler]
 */
class ApolloPrefetchCallback(callback: ApolloPrefetch.Callback, handler: Handler) : ApolloPrefetch.Callback() {
  val delegate: ApolloPrefetch.Callback
  private val handler: Handler
  override fun onSuccess() {
    handler.post { delegate.onSuccess() }
  }

  override fun onFailure(e: ApolloException) {
    handler.post { delegate.onFailure(e) }
  }

  override fun onHttpError(e: ApolloHttpException) {
    handler.post { delegate.onHttpError(e) }
  }

  override fun onNetworkError(e: ApolloNetworkException) {
    handler.post { delegate.onNetworkError(e) }
  }

  companion object {
    /**
     * Wraps `callback` to be be operated on specified `handler`
     *
     * @param callback original callback to delegates calls
     * @param handler  the callback will be run on the thread to which this handler is attached
     */
    fun <T> wrap(callback: ApolloPrefetch.Callback, handler: Handler): ApolloPrefetchCallback {
      return ApolloPrefetchCallback(callback, handler)
    }
  }

  /**
   * @param callback original callback to delegates calls
   * @param handler  the callback will be run on the thread to which this handler is attached
   */
  init {
    delegate = __checkNotNull(callback, "callback == null")
    this.handler = __checkNotNull(handler, "handler == null")
  }
}